testsuite/mlir/mlir-parse/subsumption.mlir

```diff
@@ -33,4 +33,11 @@ module {
 
     return
   }
+
+  func @test_distributivity(%a: !verona.meet<U64, join<iso, mut>>) {
```

Comment:
This is only one of many possibilities of the new code. Please add tests for at least the major variations (join of meets, meets of joins, simple types, etc).

Reply:
This (and obviously variants of it) is, I strongly believe, the only new subtyping pattern that is now recognized but wasn't before.

While the normalization code have to handle all other sorts of types, you can't really observe them just from the subtyping rules. If we want to test the normalization process in more details we should IMO do it with unit testing around just the `normalizeType` function.

Reply:
Testing for the variants is a good exercise in understanding the domain (TDD), as well as increased coverage for future changes, as well as expressing all possible cases to future developers. So I'd strongly advocate for adding them all.

Meets of joins, joins of meets, unchanging transforms (like no combination, only joins, only meets), leaving the unimplemented normalisations for later (flattening, deduplication, etc).

Reply:
What I mean is that from a subtyping POV, which is the only way normalization is invoked at the moment, for all these cases, it doesn't make a difference whether normalization was applied.

For instance, `join<A, A>` is a subtype of A, just by virtue of the subtyping rules of `join`. Similarily, `join<meet<U64, iso>, meet<U64, mut>>` is actually already a subtype of `verona.meet<U64, join<iso, mut>>`.

Therefore if we want to write any meaningful test for those, it cannot be just by testing subtyping, since the tests would have passed regardless of normalization.

Reply:
Ah, right, we're not talking about the same thing, then. :)

My point is that the normalisation code can do a lot more than this example, regardless of how it integrates with the sub-typing. I guess that's why you mentioned about unit tests, but creating a unit test infrastructure for this patch doesn't make sense.

I was just thinking one function with unrelated instructions on various types and then the result of those types to be printed back, but you don't actually change the type, just check it, so it's harder to test that way.

My main worry is that there is something in the code that is broken but we can't see because there are no tests for it. I guess for this patch, focused on sub-typing, not normalisation, we could let it be for now.

I really want to see the types normalised in the IR as soon as possible, in which case it would be easy to test. But that's for another patch. For now, looks ok.